### PR TITLE
fix(mc): handle pinning in the reducers instead of always calling refresh()

### PR DESCRIPTION
**Pull request by rlr**
@k88hudson r?

### CODE REVIEW

File: system-addon/common/Reducers.jsm [hunk 1]
```diff
@@ -51,9 +51,40 @@ function App(prevState = INITIAL_STATE.App, action) {
   }
 }
 
+function insertPinned(links, pinned) {
```
Comment (k88hudson) [inline system-addon/common/Reducers.jsm]: Can you add a jsdoc for this?

File: system-addon/common/Reducers.jsm [hunk 2]
```diff
@@ -51,9 +51,40 @@ function App(prevState = INITIAL_STATE.App, action) {
...
+  // Augment the pinned links with any other extra data we have for them already
+  let pinnedLinks = pinned.map(pinnedLink => (pinnedLink ? Object.assign(links.find(link => link && link.url === pinnedLink.url) || {}, pinnedLink, {isDefault: false}) : pinnedLink));
```
Comment (k88hudson) [inline system-addon/common/Reducers.jsm]: `const`

File: system-addon/test/unit/common/Reducers.test.js [hunk 3]
```diff
@@ -173,4 +179,69 @@ describe("Reducers", () => {
...
+      assert.equal(result[11].pinTitle, pinned[11].title);
+      assert.equal(result[11].pinIndex, 11);
+    });
+    it("should unpin previously pinned links no longer in the pinned list", () => {
+      const pinned = [];
+      links[2].isPinned = true;
+      links[2].pinTitle = "pinned site";
+      links[2].pinIndex = 2;
+      const result = insertPinned(links, pinned);
+      assert.ok(!Object.prototype.hasOwnProperty.call(result[2], "isPinned"));
```
Comment (k88hudson) [inline system-addon/test/unit/common/Reducers.test.js]: you can use `assert.notProperty`

**Comment by as-pine-proxy** [comment]
[Treeherder: c5cb128b2015a3874319a2b471f7dd7d1dd7bdbd](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=c5cb128b2015a3874319a2b471f7dd7d1dd7bdbd)

**Pull request merged**